tests/equality/equality.cpp

```diff
@@ -0,0 +1,995 @@
+/*
+Copyright libCellML Contributors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "gtest/gtest.h"
+
+#include <libcellml>
+
+#include "test_utils.h"
+
+TEST(Equality, invalidInput)
+{
+    libcellml::VariablePtr v1 = libcellml::Variable::create("variable");
+
+    EXPECT_FALSE(v1->equal(nullptr));
+}
+
+TEST(Equality, variableEqualByNameOnly)
+{
+    libcellml::VariablePtr v1 = libcellml::Variable::create("variable");
+    libcellml::VariablePtr v2 = libcellml::Variable::create("variable");
+
+    EXPECT_NE(v1, v2);
+    EXPECT_TRUE(v1->equal(v2));
+}
+
+TEST(Equality, variableNotEqualByNameOnly)
+{
+    libcellml::VariablePtr v1 = libcellml::Variable::create("variableA");
+    libcellml::VariablePtr v2 = libcellml::Variable::create("variableB");
+
+    EXPECT_NE(v1, v2);
+    EXPECT_FALSE(v1->equal(v2));
+}
+
+TEST(Equality, variableNotEqualToUnits)
+{
+    libcellml::VariablePtr v1 = libcellml::Variable::create("variable");
+    libcellml::UnitsPtr u1 = libcellml::Units::create("variable");
+
+    EXPECT_FALSE(v1->equal(u1));
+}
+
+TEST(Equality, variableEqualById)
+{
+    libcellml::VariablePtr v1 = libcellml::Variable::create("variable");
+    libcellml::VariablePtr v2 = libcellml::Variable::create("variable");
+
+    v1->setId("lefisjeij");
+    v2->setId("lefisjeij");
+
+    EXPECT_TRUE(v1->equal(v2));
+}
+
+TEST(Equality, variableNotEqualById)
+{
```

Comment:
Need to test empty id vs non-empty id too?

Reply:
I don't think so.  The test is for equal and not equal the string comparison which the id comparison is based on is well defined in this.